### ISSUE TITLE
Track artefacts on on-premise machines

### ISSUE THREAD

**Issue by dougbu** [dotnet/dnceng#769]
As discussed in the tactical #379 issue, our artefacts are not up to date on many on-premise machines. This for example means recent changes to helix-scripts/ content were unable to be used on some machines. Now, as part of `Epic: Helix Machines Improvements`, we should fund work to automate as much of this as we can. Tracking artefacts on a regular cadence will ease confirming Matrix of Truth and and https://helix.dot.net reports. It will also ease searching for machines w/ particular packages since searching the YAML can be painful.

This might involve an automated run that generates an inventory or "state of the machine", run it weekly and put it in a database, and put up a dashboard to let the state be discoverable and also highlight any differences from what's expected (as defined in the helix-machines artifacts definitions).

For bonus points, great if we can determine the exact versions of OS packages and others where we use version ranges. That will be useful for providing more specific and correct CG manifests.

This issue complements #768.

### Release Note Category

- [x] Feature changes/additions 
- [ ] Bug fixes
- [ ] Internal Infrastructure Improvements

### Release Note Description

We are now automatically tracking the configuration and installations of and on all on-premise machines.

**Comment by dougbu** [comment]
@riarenas I'm not sure whether we should also create an issue (if we don't already have one) to alert when on-prem machines fall behind on helix-scripts/ versions. what do you think❔

@ilyas1974 and @garath I have a vague idea the idea of such an alert came up a while back. did it become an issue❔

@chcosta this is one area where your on our dependencies might help us. we don't track indirect dependencies our artefacts install during image generation. can the tooling you worked with help there❔

the other area where we're a bit blind is on what actually gets installed when we specify version ranges. since that happens mostly w/ Python packages, it's likely not super relevant unless we can somehow run Component Governance scans as we generate images. is that possible❔

**Comment by garath** [comment]
> I have a vague idea the idea of such an alert came up a while back. did it become an issue❔

There's no Grafana alert tracking this sort of thing. There is a dashboard showing the count of different Helix Client versions in use [here](https://dotnet-eng-grafana.westus2.cloudapp.azure.com/d/queueOverview/on-premises-helix-queue-status?orgId=1), but no alerts associated. There is nothing in Grafana tracking artifact version. That is only tracked using the "updateBy" stamps in the helix-machines YAML.

Is that helpful?

**Comment by dougbu** [comment]
> Is that helpful?

Yes. I'm going to file yet-another Helix Machines issue about adding an alert when https://dotnet-eng-grafana.westus2.cloudapp.azure.com/d/queueOverview/on-premises-helix-queue-status?orgId=1 starts to look worse than normal outside deployment windows.

**Comment by dougbu** [comment]
found it &mdash; #660 

that issue is currently in `Ops - Service Maintenance`. any opinions @riarenas, @ilyas1974, and @garath on leaving it there versus moving it to the Helix Machines epic❔ I don't have a strong opinion

**Comment by ilyas1974** [comment]
I think this should be considered for the Helix Machines epic as you will be making changes to the service. If, as part of your changes, there is a "hard" requirement to ensure the helix scripts version is kept on "up to date", then I think we should incorporate alerting into the epic. If, on the other hand, this is something that is a best practice for the team - then let's put it into service maintenance.

**Comment by dougbu** [comment]
per your comment @ilyas1974, I moved #660 from service maintenance into the helix machines epic

**Comment by ilyas1974** [comment]
[like]  Ilya Skuratovsky reacted to your message:
________________________________
